File: cec2014/R/cec2014.R
#' @export 

cec2014 = function(i, x) {
  if (base::missing(i)) base::stop("Missing argument; 'i' has to be provided !")

  if (base::missing(x)) base::stop("Missing argument; 'x' has to be provided !")

  if (base::is.numeric(i) && i >= 1 && i <= 30) {
    if (base::is.vector(x)) {
      row = 1
      col = base::length(x)
    } else if (base::is.matrix(x)) {
      row = base::nrow(x)
      col = base::ncol(x)
    } else {
      base::stop("x should be a vector or a matrix")
    } 

    if (!(col %in% c(2, 10, 20, 30, 50, 100))) {
      base::stop("Invalid argument: Only 2, 10, 20, 30, 50, and 100 dimensions/variables are allowed !")
    }
    extdatadir = base::system.file("extdata", package = "cec2014")
    f = base::.C("cec2014",
      extdatadir = as.character(extdatadir),
      i = base::as.integer(i), x = base::as.double(x), row = base::as.integer(row),
      col = base::as.integer(col), f = double(row),
      PACKAGE = "cec2014"
    )$f
  } else {
    base::stop("Invalid argument: 'i' should be an integer between 1 and 30 !")
  }
  return(f)
}

File: rewritten/cec2014/src/cec2014.c
#include "../include/cec2014.h"
#include "../include/cec2014_interface.h"

void cec2014(char **extdatadir, int *i, double *X, int *row, int *col,
             double *f) {

  double *x = malloc(*col * sizeof(double));

  for (int r = 0; r < *row; r++) {
    for (int c = 0; c < *col; c++) {
      x[c] = X[r + *row * c];
    }
    cec2014_func(*extdatadir, x, &f[r], *col, 1, *i);
  }
  free(x);
}
